kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergCompatV3MetadataValidatorAndUpdater.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.icebergcompat;
+
+import static io.delta.kernel.internal.tablefeatures.TableFeatures.*;
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static java.util.Collections.singletonList;
+import static java.util.stream.Collectors.toList;
+
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.TableConfig;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.tablefeatures.TableFeature;
+import io.delta.kernel.internal.types.TypeWideningChecker;
+import io.delta.kernel.internal.util.ColumnMapping.ColumnMappingMode;
+import io.delta.kernel.internal.util.SchemaIterable;
+import io.delta.kernel.internal.util.SchemaUtils;
+import io.delta.kernel.internal.util.Tuple2;
+import io.delta.kernel.types.*;
+import io.delta.kernel.utils.DataFileStatus;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/** Utility methods for validation and compatibility checks for Iceberg V3. */
+public class IcebergCompatV3MetadataValidatorAndUpdater
+    extends IcebergCompatMetadataValidatorAndUpdater {
+  /**
+   * Validate and update the given Iceberg V3 metadata.
+   *
+   * @param newMetadata Metadata after the current updates
+   * @param newProtocol Protocol after the current updates
+   * @return The updated metadata if the metadata is valid and updated, otherwise empty.
+   * @throws UnsupportedOperationException if the metadata is not compatible with Iceberg V3
+   *     requirements
+   */
+  public static Optional<Metadata> validateAndUpdateIcebergCompatV3Metadata(
+      boolean isCreatingNewTable, Metadata newMetadata, Protocol newProtocol) {
+    return INSTANCE.validateAndUpdateMetadata(
+        new IcebergCompatInputContext(isCreatingNewTable, newMetadata, newProtocol));
+  }
+
+  /**
+   * Validate the given {@link DataFileStatus} that is being added as a {@code add} action to Delta
+   * Log. Currently, it checks that the statistics are not empty.
+   *
+   * @param dataFileStatus The {@link DataFileStatus} to validate.
+   */
+  public static void validateDataFileStatus(DataFileStatus dataFileStatus) {
```

Comment:
given this is common for all iceberg compact v1, lets just move this to the base class and you can still call with IcebergCompatV3MetadataValidatorAndUpdater.validateDataFileStatus